edx_proctoring/views.py

```diff
@@ -3,23 +3,177 @@
 """
 
 import logging
+from django.db import IntegrityError
+from django.db.models import Model
 
 from rest_framework import status
 from rest_framework.response import Response
+from edx_proctoring.api import create_exam, update_exam, get_exam_by_id, get_exam_by_content_id, start_exam_attempt, \
+    stop_exam_attempt, add_allowance_for_user, remove_allowance_for_user, get_active_exams_for_user
 
 from .utils import AuthenticatedAPIView
 
 LOG = logging.getLogger("edx_proctoring_views")
 
 
-class StudentProctoredExamStatus(AuthenticatedAPIView):
+class ProctoredExamView(AuthenticatedAPIView):
     """
-    Returns the status of the proctored exam.
+    Endpoint for the Proctored Exams
+    /edx_proctoring/v1/proctored_exam/exam
+
+    Supports:
+        HTTP POST: Creates a new Exam.
+        HTTP PUT: Updates an existing Exam.
+        HTTP GET: Returns an existing exam (by id or by content id)
+
+    HTTP POST
+    Creates a new Exam.
+    Expected POST data: {
+        "course_id": "edX/DemoX/Demo_Course",
+        "content_id": 123,
+        "exam_name": "Midterm",
+        "time_limit_mins": "90",
+        "is_proctored": true,
+        "external_id": "",
+        "is_active": true,
+    }
+
+    **POST data Parameters**
+        * course_id: The unique identifier for the course.
+        * content_id: This will be the pointer to the id of the piece of course_ware which is the proctored exam.
+        * exam_name: This is the display name of the Exam (Midterm etc).
+        * time_limit_mins: Time limit (in minutes) that a student can finish this exam.
+        * is_proctored: Whether this exam actually is proctored or not.
+        * external_id: This will be a integration specific ID - say to SoftwareSecure.
+        * is_active: Whether this exam will be active.
+
+
+    **Response Values**
+        * {'exam_id': ##}, The exam_id of the created Proctored Exam.
+
+    **Exceptions**
+        * HTTP_400_BAD_REQUEST, data={"message": "Trying to create a duplicate exam."}
+
+    HTTP PUT
+    Updates an existing Exam.
+    PUT data : {
+        "exam_id": 533,
+        "exam_name": "Final",
+        "time_limit_mins": 120,
+        "is_proctored": true,
+        "external_id": 235
+        "is_active": true
+    }
+
+    **PUT data Parameters**
+        see the POST data parameters
+
+    **Response Values**
+        * {'exam_id': ##}, The exam_id of the created Proctored Exam.
+
+
+    HTTP GET
+        ** Scenarios **
+        ?exam_id=533
+        returns an existing exam  object matching the exam_id
+
+        ?course_id=edX/DemoX/Demo_Course&content_id=123
+        returns an existing exam object matching the course_id and the content_id
+    """
+    def post(self, request):
```

Comment:
Just note that we have to tie in a permissions system here. Only course authors should be able to create new exams. For now, we should write simply a has_permissions(user, action) helper function and any function that would require staff permissions (e.g. create new exam, add/update/delete allowance, etc.) should check user.is_staff to see if it is on, if not raise a PermissionDenied exception.

Later on (this coming sprint) we'll have to get the real permissions checks (CourseAccessRole) but that subsystem will have to be dependency injected in from the LMS (since we don't have access to those models from this repo).

Can you apply these patterns to all POST and PUT operations? Maybe - as bonus points - you can write a method decorator?